File: python/poker/poker.py
def best_hands(hands):
    best = max(hands, key=hand_rank)
    return [
        hand
        for hand in hands
        if card_ranks(hand) == card_ranks(best)
    ]

def deserialize(hand):
    _hand = []
    for card in hand.split():
        _hand.append((card[-2], card[-1]))
    return _hand

def card_ranks(hand):
    hand = deserialize(hand)
    ranks = ["--234567890JKQA".index(r) for r, s in hand]
    ranks.sort(reverse=True)
    return [5,4,3,2,1] if ranks == [14,5,4,3,2] else ranks

def hand_rank(hand):
    ranks = card_ranks(hand)
    hand = deserialize(hand)
    groups = [(ranks.count(i), i) for i in set(ranks)]
    groups.sort(reverse=True)
    counts, descending = zip(*groups)
    flush = len(set([s for r, s in hand])) == 1
    straight = len(set([r for r, s in hand])) == 5 and (max(ranks) - min(ranks)) == 4
    return (
        8 if straight and flush else
        7 if counts == (4, 1) else
        6 if counts == (3, 2) else
        5 if flush else
        4 if straight else
        3 if counts == (3, 1, 1) else
        2 if counts == (2, 2, 1) else
        1 if counts == (2, 1, 1, 1) else
        0, descending
    )

File: ruby/poker/poker.rb
class Poker
  CARD_VALUE = { :J => 11, :Q => 12, :K => 13, :A => 14 }

  def initialize(hands)
    @hands = hands
  end

  def hand_value(hand)

    ranks = hand.map { |card| (CARD_VALUE[card[0].to_sym] || card[0..-1]).to_i }.sort
    ranks = [1,2,3,4,5] if ranks == [2,3,4,5,14]
    sets = ranks.each_with_object(Hash.new(0)) { |num, agg| agg[num] += 1 }

    flush = hand.map { |card| card[-1] }.uniq.size == 1
    straight = (ranks.max - ranks.min == 4) && (ranks.uniq.size == 5)
    twos = sets.select { |_, count| count == 2 }.keys.sort.reverse
    three, four = sets.key(3), sets.key(4)
    high = ranks.max

    {
      :royal    => (straight && flush  ? high    : 0),
      :four     => (four               ? four    : 0),
      :full     => (three && twos.any? ? [three, twos[0]] : [0, 0]),
      :flush    => (flush              ? high    : 0),
      :straight => (straight           ? high    : 0),
      :three    => (three              ? three   : 0),
      :pairs    => (twos.size == 2     ? twos    : [0, 0]),
      :pair     => (twos.any?          ? twos[0] : 0),
      :high     => ranks.reverse
    }.values
  end

  def best_hand
    hand_values = @hands.map { |hand| hand_value(hand) }
    max = hand_values.max
    @hands.select.with_index { |hand, i| hand_values[i] == max }
  end

end
